flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartition.java

```diff
@@ -0,0 +1,416 @@
+	 */
+	@VisibleForTesting
+	BoundedBlockingSubpartition(
+			int index,
+			ResultPartition parent,
+			Path filePath,
+			int maxMMapRegionSize) throws IOException {
+
+		super(index, parent);
+
+		final FileChannel fc = FileChannel.open(filePath,
+				StandardOpenOption.READ, StandardOpenOption.WRITE, StandardOpenOption.CREATE_NEW);
+
+		this.memory = new MemoryMappedBuffers(filePath, fc, maxMMapRegionSize);
+		this.readers = new HashSet<>();
+	}
+
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Checks if writing is finished.
+	 * Readers cannot be created until writing is finished, and no further writes can happen after that.
+	 */
+	public boolean isFinished() {
+		return isFinished;
+	}
+
+	@Override
+	public boolean isReleased() {
+		return isReleased;
+	}
+
+	@Override
+	public boolean add(BufferConsumer bufferConsumer) throws IOException {
+		if (isFinished()) {
+			bufferConsumer.close();
+			return false;
+		}
+
+		flushCurrentBuffer();
+		currentBuffer = bufferConsumer;
+		return true;
+	}
+
+	@Override
+	public void flush() {
+		// unfortunately, the signature of flush does not allow for any exceptions, so we
+		// need to do this discouraged pattern of runtime exception wrapping
+		try {
+			flushCurrentBuffer();
+		}
+		catch (IOException e) {
+			throw new FlinkRuntimeException(e.getMessage(), e);
+		}
+	}
+
+	private void flushCurrentBuffer() throws IOException {
+		if (currentBuffer != null) {
+			writeAndCloseBufferConsumer(currentBuffer);
+			currentBuffer = null;
+		}
+	}
+
+	private void writeAndCloseBufferConsumer(BufferConsumer bufferConsumer) throws IOException {
+		try {
+			final Buffer buffer = bufferConsumer.build();
+			try {
+				memory.writeBuffer(buffer);
+
+				numBuffersAndEventsWritten++;
+				if (buffer.isBuffer()) {
+					numDataBuffersWritten++;
+				}
+			}
+			finally {
+				buffer.recycleBuffer();
+			}
+		}
+		finally {
+			bufferConsumer.close();
+		}
+	}
+
+	@Override
+	public void finish() throws IOException {
+		synchronized (lock) {
+			checkState(!isReleased, "data partition already released");
+			checkState(!isFinished, "data partition already finished");
+
+			isFinished = true;
+			flushCurrentBuffer();
+			writeAndCloseBufferConsumer(EventSerializer.toBufferConsumer(EndOfPartitionEvent.INSTANCE));
+			memory.finishWrite();
+		}
+	}
+
+	@Override
+	public void release() throws IOException {
+		synchronized (lock) {
+			if (isReleased) {
+				return;
+			}
+
+			isReleased = true;
+			isFinished = true; // for fail fast writes
+
+			checkReaderReferencesAndDispose();
+		}
+	}
+
+	@Override
+	public ResultSubpartitionView createReadView(BufferAvailabilityListener availability) throws IOException {
+		synchronized (lock) {
+			checkState(!isReleased, "data partition already released");
+			checkState(isFinished, "writing of blocking partition not yet finished");
+
+			availability.notifyDataAvailable();
```

Comment:
`availability.notifyDataAvailable()` is better to be called as late as possible, before return reader.
Because after notification of netty stack, the netty thread would call `subpartitionView.isAvailable()` from `CreditBasedSequenceNumberingViewReader`, but this `subpartitionView` might not be returned to `CreditBasedSequenceNumberingViewReader` at that moment.

Reply:
I see the point, but this seems like making assumptions about a specific user of this class, which one should usually avoid.

My understanding is that on the netty side, this should go through the same event loop thread, so the `notifyDataAvailable()` enqueues a Runnable into the event loop, which should be fine. Is that correct?

Reply:
Yes, it is true. `notifyDataAvailable` and `createReadView` would be from the same netty thread. So we could ignore this issue.